src/winmain.cpp

```diff
@@ -36,6 +37,20 @@
 #define CURL_STATICLIB
 #include "../curl/include/curl/curl.h"
 
+#define GUP_LOG_FILENAME L"c:\\tmp\\winup.log"
+
+#ifdef _DEBUG
+#define WRITE_LOG(fn, suffix, log) writeLog(fn, suffix, log);
+#else
+#define WRITE_LOG(fn, suffix, log)
+#endif
+
+// when defined, GUP will try to wait for the maintained app exit before proceeding to a scheduled op
+#define MAINTAINED_APP_RUNNING_BEHIND_TRY_TO_WAIT
+// when undefined, GUP ops will try to proceed even after an unsuccessful waiting for the maintained app exit
+// (there are also ops for which it is absolutely ok)
+//#define MAINTAINED_APP_RUNNING_BEHIND_DO_NOT_OPERATE

```

Comment:
The above 2 MACRO are for debug use?

Reply:
PR is finished. It was good for me to write it in this way (to be able to test many possible variants), so I left it as it is now.

If you like (?), I can of course remove both macros (MAINTAINED_APP_RUNNING_BEHIND_TRY_TO_WAIT and MAINTAINED_APP_RUNNING_BEHIND_DO_NOT_OPERATE) and still let the code run as it is now.

Reply:
> If you like (?), I can of course remove both macros (MAINTAINED_APP_RUNNING_BEHIND_TRY_TO_WAIT and MAINTAINED_APP_RUNNING_BEHIND_DO_NOT_OPERATE) and still let the code run as it is now.

Please remove them to make the code more readable.

Also though I didn't see any dead lock case, please ensure that will never happen - Indeed user could have the 2nd Notepad++ opened but hidden - in this case if we wait infinitely for "nppInstance" mutex, dead lock will happen.

Reply:
@donho 

> Please remove them to make the code more readable.

OK

> user could have the 2nd Notepad++ opened but hidden - in this case if we wait infinitely for "nppInstance" mutex, dead lock will happen.

No, that's impossible with my PR - there is max waiting time (5s), then wingup tries to continue, e.g. see the above log-example:

> Waiting for the app-exit failed: The maintained app seems to be still running but GUP will try to proceed anyway...